dikastes.go

```diff
@@ -1,44 +1,163 @@
+  -l --listen <port>     Unix domain socket path [default: /var/run/dikastes/dikastes.sock]
+  -d --dial <target>     Target to dial. [default: localhost:50051]
+  --debug                Log at Debug level.`
 const version = "0.1"
-const default_port = ":50051"
-
+const NODE_NAME_ENV = "K8S_NODENAME"
 
 func main() {
 	arguments, err := docopt.Parse(usage, nil, true, version, false)
 	if err != nil {
 		println(usage)
 		return
 	}
+	if arguments["--debug"].(bool) {
+		log.SetLevel(log.DebugLevel)
+	}
 	if arguments["server"].(bool) {
-		lis, err := net.Listen("tcp", default_port)
+		runServer(arguments)
+	} else if arguments["client"].(bool) {
+		runClient(arguments)
+	}
+	return
+}
+
+func runServer(arguments map[string]interface{}) {
+	filePath := arguments["--listen"].(string)
+	_, err := os.Stat(filePath)
+	if !os.IsNotExist(err) {
+		// file exists, try to delete it.
+		err := os.Remove(filePath)
 		if err != nil {
-			log.Fatalf("Unable to listen on %v")
+			log.WithFields(log.Fields{
+				"listen": filePath,
+				"err":    err,
+			}).Fatal("File exists and unable to remove.")
 		}
-		s := grpc.NewServer()
-		authz.RegisterAuthorizationServer(s, server.NewServer())
-		reflection.Register(s)
-		if err := s.Serve(lis); err != nil {
+	}
+	lis, err := net.Listen("unix", filePath)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"listen": filePath,
+			"err":    err,
+		}).Fatal("Unable to listen.")
+	}
+	defer lis.Close()
+	err = os.Chmod(filePath, 0777) // Anyone on system can connect.
+	if err != nil {
+		log.Fatal("Unable to set write permission on socket.")
+	}
+	gs := grpc.NewServer()
+	store := policystore.NewPolicyStore()
+	checkServer := checker.NewServer(store)
+	authz.RegisterAuthorizationServer(gs, checkServer)
+
+	// Synchronize the policy store
+	opts := getDialOptions()
+	syncClient := syncher.NewClient(arguments["--dial"].(string), opts)
+	syncContext, cancelSync := context.WithCancel(context.Background())
+	defer cancelSync()
+	go syncClient.Sync(syncContext, store)
+
+	// Run gRPC server on separate goroutine so we catch any signals and clean up.
+	go func() {
+		if err := gs.Serve(lis); err != nil {
 			log.Fatalf("failed to serve: %v", err)
 		}
+	}()
+
+	// Use a buffered channel so we don't miss any signals
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, os.Kill, syscall.SIGTERM)
+
+	// Block until a signal is received.
+	s := <-c
+	log.Infof("Got signal:", s)
+}
+
+func runClient(arguments map[string]interface{}) {
+	opts := getDialOptions()
+	conn, err := grpc.Dial(arguments["--dial"].(string), opts...)
+	if err != nil {
+		log.Fatalf("fail to dial: %v", err)
+	}
+	defer conn.Close()
+	client := authz.NewAuthorizationClient(conn)
+	req := authz.CheckRequest{
+		Attributes: &authz.AttributeContext{
+			Source: &authz.AttributeContext_Peer{
+				Principal: fmt.Sprintf("spiffe://cluster.local/ns/%s/sa/%s",
+					arguments["<namespace>"].(string), arguments["<account>"].(string)),
```

Comment:
I think it might read better if we parse all the arguments in the beginning. Like we do here: https://github.com/projectcalico/calicoctl/blob/master/calicoctl/resourcemgr/resourcemgr.go#L287-L289

You might also find the `argutils` package from calicoctl useful

Reply:
+1